porcupine/actions.py

```diff
@@ -0,0 +1,115 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Callable, Union
+
+from porcupine.tabs import FileTab
+
+action_availability_callback = Callable[[], bool]
+
+
+@dataclass(frozen=True)
+class BareAction:
+    """Action that requires no context in the callback"""
+
+    name: str
+    description: str
+    callback: Callable[[], None]
+    availability_callback: action_availability_callback
+
+
+filetab_action_availability_callback = Callable[[FileTab], bool]
+
+
+@dataclass(frozen=True)
+class FileTabAction:
+    """Action that requires a FileTab to be provided to the callback"""
+
+    name: str
+    description: str
+    callback: Callable[[FileTab], None]
+    availability_callback: filetab_action_availability_callback
+
+
+path_action_availability_callback = Callable[[Path], bool]
+
+
+@dataclass(frozen=True)
+class PathAction:
+    """Action that requires a Path to be provided to the callback"""
+
+    name: str
+    description: str
+    callback: Callable[[Path], None]
+    availability_callback: path_action_availability_callback
+
+
+Action = Union[BareAction, FileTabAction, PathAction]
+
+_actions: dict[str, Action] = {}
+
+
+def register_action(
+    *,
+    name: str,
+    description: str,
+    callback: Callable[..., None],
+    availability_callback: action_availability_callback = lambda: True,
+) -> BareAction:
+    if name in _actions:
+        raise ValueError(f"Action with the name '{name}' already exists")
+    action = BareAction(
+        name=name,
+        description=description,
+        callback=callback,
+        availability_callback=availability_callback,
+    )
+    _actions[name] = action
+    return action
+
+
+def register_filetab_action(
+    *,
+    name: str,
+    description: str,
+    callback: Callable[[FileTab], None],
+    availability_callback: filetab_action_availability_callback = lambda tab: True,
+) -> FileTabAction:
+    if name in _actions:
+        raise ValueError(f"Action with the name '{name}' already exists")
+    action = FileTabAction(
+        name=name,
+        description=description,
+        callback=callback,
+        availability_callback=availability_callback,
+    )
+    _actions[name] = action
+    return action
+
+
+def register_path_action(
+    *,
+    name: str,
+    description: str,
+    callback: Callable[[Path], None],
+    availability_callback: path_action_availability_callback = lambda path: True,
+) -> PathAction:
+    if name in _actions:
+        raise ValueError(f"Action with the name '{name}' already exists")
+    action = PathAction(
+        name=name,
+        description=description,
+        callback=callback,
+        availability_callback=availability_callback,
+    )
+    _actions[name] = action
+    return action
+
+
+def query_actions(name: str) -> Action | None:
```

Comment:
```suggestion
def find_action(name: str) -> Action | None:
```

Reply:
Was this resolved?

Reply:
@benjamin-kirkbride ?

Reply:
Sorry, somehow missed this.

How about `get_actions` since it can return None (and in fact is just a thin wrapper around `dict.get`)?

Reply:
Fine for me.

Reply:
But `get_action` then. So not plural.

Reply:
Yes, of course :sweat_smile:

Reply:
It looks like a consensus has been reached 😄.